Nil pointer dereference in Transport.DoRaw when connecting to TLS-enabled Docker daemon

Just cross-posting this issue here from Copa, were experiencing a nil pointer dereference in `transport.(*Transport).DoRaw` when attempting to connect to a TLS-enabled Docker daemon.

Stack trace:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xc5ff54]
goroutine 47 [running]:
[github.com/cpuguy83/go-docker/transport.(*Transport).DoRaw(0xc000c50280](https://github.com/cpuguy83/go-docker/transport.(*Transport).DoRaw(0xc000c50280), {0x4184f88, 0xc000790300}, {0x35103ff, 0x4}, {0xc000c0a068, 0x5}, {0xc0009f4008, 0x1, 0x1})
/home/runner/go/pkg/mod/[github.com/cpuguy83/go-docker@v0.3.0/transport/transport.go:84](https://github.com/cpuguy83/go-docker@v0.3.0/transport/transport.go:84) +0x2f4
[github.com/project-copacetic/copacetic/pkg/buildkit.autoClient.Docker.func2](https://github.com/project-copacetic/copacetic/pkg/buildkit.autoClient.Docker.func2)({0x4184f88, 0xc000790300}, {0x413e110?, 0x35ca134?})
/home/runner/work/copacetic/copacetic/pkg/buildkit/connhelpers/docker.go:28 +0x151
google.golang.org/grpc/internal/transport.dial({0x4184f88, 0xc000790300}, 0xc0000944c0, {{0x35ca134, 0x1c}, {0x3529441, 0x9}, 0xc0000c4618, 0x0, {0x0, ...}}, ...)
/home/runner/go/pkg/mod/google.golang.org/grpc@v1.70.0/internal/transport/http2_client.go:171 +0x172
... (rest of stack trace)
```

I believe This is the relevant line in copacetic: https://github.com/project-copacetic/copacetic/blob/d134be64d25ca7ecd841e7094d13b3e47a7bbb39/pkg/buildkit/connhelpers/docker.go#L36

Any insights would be much appreciated!

--- a/transport/tcp.go
+++ b/transport/tcp.go
@@ -1,6 +1,7 @@
 package transport
 
 import (
+	"context"
 	"net"
 	"net/http"
 )
@@ -14,19 +15,27 @@ func TCPTransport(host string, opts ...ConnectionOption) (*Transport, error) {
 		}
 	}
 
-	t := &Transport{
-		scheme: "http",
-		host:   host,
-		c: &http.Client{
-			Transport: &http.Transport{
-				DialContext:     new(net.Dialer).DialContext,
-				TLSClientConfig: cfg.TLSConfig,
-			},
-		},
+	httpTransport := &http.Transport{
+		DialContext:     new(net.Dialer).DialContext,
+		TLSClientConfig: cfg.TLSConfig,
 	}
 
+	scheme := "http"
 	if cfg.TLSConfig != nil {
-		t.scheme = "https"
+		scheme = "https"
+	}
+
+	dial := func(ctx context.Context) (net.Conn, error) {
+		return httpTransport.DialContext(ctx, "tcp", host)
+	}
+
+	t := &Transport{
+		scheme: scheme,
+		host:   host,
+		c: &http.Client{
+			Transport: httpTransport,
+		},
+		dial: dial,
 	}
 
 	return t, nil

--- a/transport/tcp_test.go
+++ b/transport/tcp_test.go
@@ -2,6 +2,7 @@ package transport
 
 import (
 	"context"
+	"crypto/tls"
 	"io"
 	"net/http"
 	"net/http/httptest"
@@ -34,4 +35,48 @@ func TestTCPTransport(t *testing.T) {
 	_, err = io.ReadFull(resp.Body, buf)
 	assert.NilError(t, err)
 	assert.Equal(t, string(buf), data)
+}
+
+func TestTCPTransportDoRaw(t *testing.T) {
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// for DoRaw test, we expect an upgrade request
+		if req.Header.Get("Connection") == "Upgrade" {
+			w.WriteHeader(http.StatusSwitchingProtocols)
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+		}
+	}))
+	defer srv.Close()
+
+	ctx := context.Background()
+
+	u, err := url.Parse(srv.URL)
+	assert.NilError(t, err)
+
+	tr, err := TCPTransport(u.Host)
+	assert.NilError(t, err)
+
+	conn, err := tr.DoRaw(ctx, "POST", "/grpc", WithUpgrade("h2c"))
+	assert.Assert(t, conn != nil, "expected a connection but got nil")
+	assert.NilError(t, err)
+}
+
+func TestTCPTransportDoRawWithTLS(t *testing.T) {
+	// test with TLS configuration
+	tlsOpt := func(cfg *ConnectionConfig) error {
+		cfg.TLSConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+		return nil
+	}
+
+	tr, err := TCPTransport("localhost:2376", tlsOpt)
+	assert.NilError(t, err)
+
+	ctx := context.Background()
+
+	_, err = tr.DoRaw(ctx, "POST", "/grpc", WithUpgrade("h2c"))
+	// we expect this to fail with connection error since theres no server
+	// but it should not panic
+	assert.Assert(t, err != nil, "expected an error but got none")
 }

